modules/top-level/test.nix

```diff
@@ -1,10 +1,90 @@
-{ lib, ... }:
+{
+  pkgs,
+  config,
+  lib,
+  ...
+}:
+let
+  cfg = config.test;
+in
 {
   options.test = {
+    name = lib.mkOption {
+      type = lib.types.str;
+      default = "nixvim-check";
+      description = "The test derivation's name.";
+    };
+
     runNvim = lib.mkOption {
       type = lib.types.bool;
       description = "Whether to run `nvim` in the test.";
       default = true;
     };
+
+    # Output
+    derivation = lib.mkOption {
+      type = lib.types.package;
+      description = ''
+        A derivation that tests the config by running neovim.
+      '';
+      readOnly = true;
+    };
+  };
+
+  config = {
+    test.derivation = pkgs.stdenv.mkDerivation {
+      inherit (cfg) name;
+      dontUnpack = true;
+
+      nativeBuildInputs = [
+        config.finalPackage
+        # TODO: Do we really need docker?
+        pkgs.docker-client
+      ];
+
+      # First check warnings/assertions, then run nvim
+      buildPhase =
+        let
+          inherit (config) warnings;
+          assertions = lib.nixvim.modules.getAssertionMessages config.assertions;
+
+          showErr =
+            name: lines:
+            lib.optionalString (lines != [ ]) ''
+              Unexpected ${name}:
+              ${lib.concatStringsSep "\n" (lib.map (v: "- ${v}") lines)}
+            '';
+
+          errors = lib.foldlAttrs (
+            err: name: lines:
+            err + showErr name lines
+          ) "" { inherit assertions warnings; };
+        in
+        ''
+          ${lib.toShellVars { inherit errors; }}
+          if [ -n "$errors" ]; then
+            echo -n "$errors"
+            exit 1
+          fi
+        ''
+        # We need to set HOME because neovim will try to create some files
+        #
+        # Because neovim does not return an exitcode when quitting we need to check if there are
+        # errors on stderr
+        + lib.optionalString cfg.runNvim ''
+          mkdir -p .cache/nvim
+
+          output=$(HOME=$(realpath .) nvim -mn --headless "+q" 2>&1 >/dev/null)
+          if [[ -n $output ]]; then
+            echo "ERROR: $output"
+            exit 1
+          fi
+        '';
+
+      # If we don't do this nix is not happy
+      installPhase = ''
+        touch $out
```

Comment:
I was curious about this recently, it seems like we generate the empty out directory for tests. Is there a way we can output the generated config here?

Reply:
The reason we do `mkdir $out` or `touch $out` in tests is that nix considers `$out` not existing to be a build failure.

What we can do instead to make the config (or anything else) available is add extra attrs to the derivation, since a derivation is represented in nix as an attrset.

The idiomatic way to do this would be to define a `passthru` attr

e.g. `passthru.config = result.config`

But I figured this PR was way too big already, without also adding passthru attrs to our tests!